ledger/simulation/tracer.go

```diff
@@ -316,7 +326,9 @@ func (tracer *evalTracer) AfterOpcode(cx *logic.EvalContext, evalError error) {
 		}
 
 		latestOpcodeTraceUnit := &(*txnTrace.programTraceRef)[len(*txnTrace.programTraceRef)-1]
-		latestOpcodeTraceUnit.appendAddedStackValue(cx, tracer)
+		if tracer.result.ReturnStackChange() {
+			latestOpcodeTraceUnit.appendAddedStackValue(cx, tracer)
+		}
```

Comment:
Why is the no similar block like:
```
		if tracer.result.ReturnScratchChange() {
			latestOpcodeTraceUnit.somethingAboutScratch...
		}
```

Reply:
For general opcodes, there exists a symmetry that: before opcode we record what's going to be removed from stack, after opcode we record what's going to be added to the stack.

While for `store` or `stores`, we can infer the write to the scratch slot when we see these opcodes. But this is a good reminder -- I should have put the snippet of code for scratch change in *after opcode*, for if the opcode is errored, it makes no sense to record the not happening change in exec trace, see [record scratch change only after opcode](https://github.com/algorand/go-algorand/pull/5563/commits/61d351fee5e08350b49b6ece09061e1348349352)

Reply:
I made a symmetry in [symmetry in before and after opcode](https://github.com/algorand/go-algorand/pull/5563/commits/e1f51d69a7ce50c22c1832a0268dc54f2a020a7f)